rust/cyclotron-core/src/worker.rs

```diff
@@ -0,0 +1,245 @@
+    // radically simplifies using this for FFI (because there's no message passing across runtimes)
+    pending: Arc<Mutex<HashMap<Uuid, JobUpdate>>>,
+}
+
+impl Worker {
+    pub async fn new(config: PoolConfig) -> Result<Self, QueueError> {
+        let pool = config.connect().await?;
+        Ok(Self {
+            pool,
+            pending: Arc::new(Mutex::new(HashMap::new())),
+        })
+    }
+
+    pub fn from_pool(pool: PgPool) -> Self {
+        Self {
+            pool,
+            pending: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+
+    /// Dequeues jobs from the queue, and returns them. Job sorting happens at the queue level,
+    /// workers can't provide any filtering or sorting criteria - queue managers decide which jobs are run,
+    /// workers just run them.
+    pub async fn dequeue_jobs(
+        &self,
+        queue: &str,
+        worker_type: WaitingOn,
+        limit: usize,
+    ) -> Result<Vec<Job>, QueueError> {
+        let jobs = dequeue_jobs(&self.pool, queue, worker_type, limit).await?;
+
+        let mut pending = self.pending.lock().unwrap();
+        for job in &jobs {
+            // We need to hang onto the locks for a job until we flush it, so we can send updates.
+            let update = JobUpdate::new(
+                job.lock_id
+                    .expect("Yell at oliver that the dequeuing code is broken. He's very sorry that your process just panicked"),
+            );
+            pending.insert(job.id, update);
+        }
+
+        Ok(jobs)
+    }
+
+    /// This is the same as dequeue_jobs, but it also returns the vm_state of the job
+    pub async fn dequeue_with_vm_state(
+        &self,
+        queue: &str,
+        worker_type: WaitingOn,
+        limit: usize,
+    ) -> Result<Vec<Job>, QueueError> {
+        let jobs = dequeue_with_vm_state(&self.pool, queue, worker_type, limit).await?;
+
+        let mut pending = self.pending.lock().unwrap();
+        for job in &jobs {
+            // We need to hang onto the locks for a job until we flush it, so we can send updates.
+            let update = JobUpdate::new(
+                job.lock_id
+                    .expect("Yell at oliver that the dequeuing (with vm) code is broken. He's very sorry that your process just panicked"),
+            );
+            pending.insert(job.id, update);
+        }
+
+        Ok(jobs)
+    }
+
+    /// NOTE - This function can only be called once, even though the underlying
+    /// basic operation can be performed as many times as the caller likes (so long as
+    /// the job state is never set to something other than running, as that clears the
+    /// job lock). We're more strict here (flushes can only happen once, you must
+    /// flush some non-running state) to try and enforce a good interaction
+    /// pattern with the queue. I might return to this and loosen this constraint in the
+    /// future, if there's a motivating case for needing to flush partial job updates.
```

Comment:
Kind of like above, I'm still confused why our whole update API isn't just a "I'm done with the job, here is its new state" type call.